doc/kernel/usermode/syscalls.rst

```diff
@@ -0,0 +1,259 @@
+  passed in have been validated if it was invoked from user mode.
+
+* A **handler function**, which wraps the implementation function and does
+  validation of all the arguments passed in.
+
+C Prototype
+===========
+
+The C prototype represents how the API is invoked from either user or
+supervisor mode. For example, to initialize a semaphore:
+
+.. code-block:: c
+
+    __syscall void k_sem_init(struct k_sem *sem, unsigned int initial_count,
+                              unsigned int limit);
+
+The :c:macro:`__syscall` attribute is very special. To the C compiler, it
+simply expands to 'static inline'. However to the post-build
+``gen_syscalls.py`` script, it indicates that this API is a system call and
+generates the body of thefunction.  The gen_syscalls.py script does some
+parsing of the function prototype, to determine the data types of its return
+value and arguments, and has some limitations:
+
+* Array arguments must be passed in as pointers, not arrays. For example,
+  ``int foo[]`` or ``int foo[12]`` is not allowed, but should instead be
+  expressed as ``int *foo``.
+
+* Function pointers horribly confuse the limited parser. The workaround is
+  to typedef them first, and then express in the argument list in terms
+  of that typedef.
+
+* :c:macro:`__syscall` must be the first thing in the prototype and cannot
+  be preceded by anything else.
+
+Any header file that declares system calls must include a special generated
+header at the very bottom on the header file. This header follows the
+naming convention ``syscalls/<name of header file>``. For example, at the
+bottom of ``include/sensor.h``:
+
+.. code-block:: c
+
+    #include <syscalls/sensor.h>
+
+Implementation Details
+----------------------
+
+Declaring an API with :c:macro:`__syscall` causes some code to be generated in
+C and header files, all of which can be found in the project out directory
+under ``include/generated/``:
+
+* The system call is added to the enumerated type of system call IDs,
+  which is expressed in include/generated/syscall_list.h. It is the name
+  of the API in uppercase, prefixed with ``K_SYSCALL_``.
+
+* A prototype for the handler function is also created in
+  ``include/generated/syscall_list.h``
+
+* An entry for the system call is created in the dispatch table
+  ``_k_sycall_table``, expressed in ``include/generated/syscall_dispatch.c``
+
+* A weak handler function is declared, which is just an alias of the
+  'unimplemented system call' handler. This is necessary since the real
+  handler function may or may not be built depending on the kernel
+  configuration. For example, if a user thread makes a sensor subsystem
+  API call, but the sensor subsystem is not enabled, the weak handler
+  will be invoked instead.
+
+The body of the API is created in the generated system header. Using the
+example of :c:func:`k_sem_init()`, this API is declared in
+``include/kernel.h``. At the bottom of ``include/kernel.h`` is:
+
+::
```

Comment:
convention is to put the double colon at the end of the sentence to start a code block:

     ``include/kernel.h``. At the bottom of ``include/kernel.h`` is::

        #include <syscalls/kernel.h>